launcher/ui/dialogs/skins/SkinManageDialog.cpp

```diff
@@ -0,0 +1,477 @@
+        QFile::rename(path, path + ".png");
+    }
+}
+
+class WaitTask : public Task {
+   public:
+    WaitTask() : m_loop(), m_done(false){};
+    virtual ~WaitTask() = default;
+
+   public slots:
+    void quit()
+    {
+        m_done = true;
+        m_loop.quit();
+    }
+
+   protected:
+    virtual void executeTask()
+    {
+        if (!m_done)
+            m_loop.exec();
+        emitSucceeded();
+    };
+
+   private:
+    QEventLoop m_loop;
+    bool m_done;
+};
+
+void SkinManageDialog::on_userBtn_clicked()
+{
+    auto user = ui->urlLine->text();
+    if (user.isEmpty()) {
+        return;
+    }
+    MinecraftProfile mcProfile;
+    auto path = FS::PathCombine(m_list.getDir(), user + ".png");
+
+    NetJob::Ptr job{ new NetJob(tr("Download user skin"), APPLICATION->network(), 1) };
+    job->setAskRetry(false);
+
+    auto uuidOut = std::make_shared<QByteArray>();
+    auto profileOut = std::make_shared<QByteArray>();
+
+    auto uuidLoop = makeShared<WaitTask>();
+    auto profileLoop = makeShared<WaitTask>();
+
+    auto getUUID = Net::Download::makeByteArray("https://api.mojang.com/users/profiles/minecraft/" + user, uuidOut);
+    auto getProfile = Net::Download::makeByteArray(QUrl(), profileOut);
+    auto downloadSkin = Net::Download::makeFile(QUrl(), path);
+
+    connect(getUUID.get(), &Task::aborted, uuidLoop.get(), &WaitTask::quit);
+    connect(getUUID.get(), &Task::failed, uuidLoop.get(), &WaitTask::quit);
+    connect(getProfile.get(), &Task::aborted, profileLoop.get(), &WaitTask::quit);
+    connect(getProfile.get(), &Task::failed, profileLoop.get(), &WaitTask::quit);
+
+    connect(getUUID.get(), &Task::succeeded, this, [uuidLoop, uuidOut, job, getProfile] {
+        try {
+            QJsonParseError parse_error{};
+            QJsonDocument doc = QJsonDocument::fromJson(*uuidOut, &parse_error);
+            if (parse_error.error != QJsonParseError::NoError) {
+                qWarning() << "Error while parsing JSON response from Minecraft skin service at " << parse_error.offset
+                           << " reason: " << parse_error.errorString();
+                uuidLoop->quit();
+                return;
+            }
+            const auto root = doc.object();
+            auto id = Json::ensureString(root, "id");
+            if (!id.isEmpty()) {
+                getProfile->setUrl("https://sessionserver.mojang.com/session/minecraft/profile/" + id);
+            } else {
+                job->abort();
+            }
+        } catch (const Exception& e) {
+            qCritical() << "Couldn't load skin json:" << e.cause();
+        }
+        uuidLoop->quit();
+    });
+
+    connect(getProfile.get(), &Task::succeeded, this, [profileLoop, profileOut, job, getProfile, &mcProfile, downloadSkin] {
+        if (Parsers::parseMinecraftProfileMojang(*profileOut, mcProfile)) {
+            downloadSkin->setUrl(mcProfile.skin.url);
+        } else {
+            job->abort();
```

Comment:
should this branch report the error to the user?

Reply:
No, as I consider all steps a whole task here: I display a generic fail message at the end.

Reply:
I take it you mean the error if the resulting skin after all the tasks end isn't valid?

Reply:
yes: if the task fails in any way the skin is not downloaded to path making it invalid